### PR TITLE
bpo-40939: Remove the old parser

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue40939

### CODE REVIEW

**Review by remilapeyre — COMMENTED** [review]
Hi, there is also a note in `Doc/using/cmdline.rst` that will need to be removed.

**Comment by pablogsal** [comment]
> Hi, there is also a note in `Doc/using/cmdline.rst` that will need to be removed.

Thanks for the catch!

**Review by lysnikolaou — APPROVED** [review]
This LGTM! Amazing work, I couldn't find a single miss! 🚀

Also, I guess this is what we've been really working for in the last months. Feels good!

**Comment by pablogsal** [comment]
I plan to land this tomorrow as this PR keeps getting merge conflicts from everywhere and it will be easier to fix small things in future PRs if needed.

File: Lib/test/test_peg_parser.py [hunk 1]
```diff
@@ -1,803 +0,0 @@
-import ast
```
Comment (gvanrossum) [inline Lib/test/test_peg_parser.py:1]: why is this file deleted?

File: Lib/test/test_string_literals.py [hunk 2]
```diff
@@ -168,8 +167,6 @@ def test_eval_bytes_invalid_escape(self):
             eval("b'''\n\\z'''")
         self.assertEqual(len(w), 1)
         self.assertEqual(w[0].filename, '<string>')
-        if use_old_parser():
```
Comment (gvanrossum) [inline Lib/test/test_string_literals.py:171]: Isn't this about something the peg parser doesn't do yet?

File: Makefile.pre.in [hunk 3]
```diff
@@ -300,32 +300,24 @@ LIBFFI_INCLUDEDIR=	@LIBFFI_INCLUDEDIR@
 # Parser
 
 PEGEN_OBJS=		\
-		Parser/pegen/pegen.o \
-		Parser/pegen/parse.o \
-		Parser/pegen/parse_string.o \
-		Parser/pegen/peg_api.o
+		Parser/pegen.o \
+		Parser/parse.o \
+		Parser/parse_string.o \
```
Comment (gvanrossum) [inline Makefile.pre.in]: Can we,rename this? The common start with pegen.c always,slows down Emacs filename completion.

File: Makefile.pre.in [hunk 4]
```diff
@@ -893,15 +873,6 @@ regen-token:
 		$(srcdir)/Grammar/Tokens \
 		$(srcdir)/Lib/token.py
 
-.PHONY: regen-keyword
-regen-keyword:
```
Comment (gvanrossum) [inline Makefile.pre.in]: So how is keyword.py kept up to date?

File: Modules/Setup [hunk 5]
```diff
@@ -134,9 +134,6 @@ faulthandler faulthandler.c
 # can call _PyTraceMalloc_NewReference().
 _tracemalloc _tracemalloc.c
 
-# PEG-based parser module -- slated to be *the* parser
-_peg_parser _peg_parser.c
```
Comment (gvanrossum) [inline Modules/Setup:138]: Why delete this? It serves a purpose for our timing tools.

File: Lib/test/test_peg_parser.py (hunk 1 above)
Comment (pablogsal) [reply Lib/test/test_peg_parser.py:1]: This file is comparing the ast of the old parser to the last of the new parser, so it serves no purpose anymore. We have been adding all negative cases (testing extensions) into other files so we won't lose anything

File: Makefile.pre.in (hunk 3 above)
Comment (lysnikolaou) [reply Makefile.pre.in]: What about `string_parser.*`?

File: Modules/Setup (hunk 5 above)
Comment (pablogsal) [reply Modules/Setup:138]: IIRC, we talked about removing this with the old parser as its main purpose was to be able to interface with both parsers (with some extra options), I think it will be simpler if we just adapt our scripts (in a different PR) to just use `ast.parse` or similar for timings of the new parser. The overhead of constructing the AST will be common on every run, so you could do still the same things we could do before. This way we have less complexity, fewer files to compile and less moving pieces

File: Makefile.pre.in (hunk 4 above)
Comment (lysnikolaou) [reply Makefile.pre.in]: I think it'd be better to keep this target and support keyword generation in the pegen parser generator, so that it's able to create `keyword.py`. I could do that today. @pablogsal Do you agree and if so, could I just push to your branch?

Comment (pablogsal) [reply Makefile.pre.in]: I was planning to do another PR in this issue with a new file in pegen that generates the keywords but I didn't want to add this to this PR because is already massive and is easier to review the less "new lines" in the diff it has

Comment (lysnikolaou) [reply Makefile.pre.in]: Do you already have something or should I do it?

File: Makefile.pre.in (hunk 3 above)
Comment (pablogsal) [reply Makefile.pre.in]: Do you want to rename `parse.c` as well?

File: Makefile.pre.in (hunk 4 above)
Comment (pablogsal) [reply Makefile.pre.in]: > Do you already have something or should I do it?

I have started something already, but feel to create a PR if you want :)

Use the same issue as this PR is using, please

File: Lib/test/test_string_literals.py (hunk 2 above)
Comment (pablogsal) [reply Lib/test/test_string_literals.py:171]: Actually, the new parser handles this correctly. I have removed the guards and bring back the checks in 0c61d04

File: Makefile.pre.in (hunk 4 above)
Comment (pablogsal) [reply Makefile.pre.in]: Cross-referencing the PR here: https://github.com/python/cpython/pull/20800

File: Makefile.pre.in (hunk 3 above)
Comment (pablogsal) [reply Makefile.pre.in]: What about `pegen.*`, `parser.c`, `string_parser.*`?

Comment (gvanrossum) [reply Makefile.pre.in]: Yes please.

Comment (pablogsal) [reply Makefile.pre.in]: Done in e11e688

**Comment by pablogsal** [comment]
I have rebased again to solve more conflicts. Could you review again?

File: Modules/Setup (hunk 5 above)
Comment (gvanrossum) [reply Modules/Setup:138]: Okay.

**Review by gvanrossum — APPROVED** [review]
I wonder if we could backport the file renames? (Not the moves from Parser/pegen/* to Parser/*, but the rename of parse.c to parser.c and parse_string.c to string_parser.c.)

**Comment by pablogsal** [comment]
> I wonder if we could backport the file renames? (Not the moves from Parser/pegen/* to Parser/*, but the rename of parse.c to parser.c and parse_string.c to string_parser.c.)

👍 I will try to make a PR later today with that

**Comment by gvanrossum** [comment]
Land it!
